### ISSUE TITLE
[DX] Add completion to Symfony commands

### ISSUE THREAD

**Issue by GromNaN** [symfony/symfony#43594]
Completion of command arguments and options is coming in Symfony 5.4 (#38275). Command and option names are already autocompleted, but argument and option values needs to be implemented on each command.

As an example, [#42251](https://github.com/symfony/symfony/pull/42251/files#diff-771e524eaa21e31d32047fda540ba0b53222c74e693bee5a45f6da7fa1644571R87-R96) added completion to  `secrets:remove`.

This feature needs to be implemented on Symfony commands that has predictable values for their arguments and options.

Some guidelines:

- Don't complete file/directory paths for now. See https://github.com/symfony/symfony/issues/43594#issuecomment-947385485
- You don't need to filter the suggestions in PHP, `compgen` (used by the shell script) does this for you based on the suggestions provided by PHP
- You can use the debugging utility while working on providing suggestions. See https://github.com/symfony/symfony/issues/43594#issuecomment-947535113

**PLEASE, comment here BEFORE starting to work on something and WAIT for the confirmation to avoid duplicate work.**

*All commands have been claimed*

 * [x] Command `debug:event-dispatcher` (`event`, `--dispatcher`, `--format`) @stephenkhoo https://github.com/symfony/symfony/pull/43859
 * [x] Command `debug:config` (`name`, `path`) @eclairia https://github.com/symfony/symfony/pull/43850
 * [x] Command `debug:autowiring` (`search`) @eclairia https://github.com/symfony/symfony/pull/43838
 * [x] Command `messenger:failed:remove` (`id`, `--transport`) @scyzoryck https://github.com/symfony/symfony/pull/43663
 * [x] Command `messenger:failed:retry` (`id`, `--transport`) @scyzoryck https://github.com/symfony/symfony/pull/43663
 * [x] Command `messenger:failed:show` (`id`, `--transport`) @scyzoryck https://github.com/symfony/symfony/pull/43663
 * [x] Command `debug:container` (`name`) @GromNaN https://github.com/symfony/symfony/pull/43857
 * [x] Command `messenger:consume` (`receivers`, `--bus`, `--queues`) @GromNaN https://github.com/symfony/symfony/pull/43891
 * [x] Command `debug:firewall` (`name`, `--format`) @IonBazan https://github.com/symfony/symfony/pull/43598
 * [x] Command `debug:form` (`name`, `--format`) @IonBazan https://github.com/symfony/symfony/pull/43598
 * [x] Command `debug:messenger` (`bus`) @IonBazan https://github.com/symfony/symfony/pull/43598
 * [x] Command `debug:router` (`name`) @IonBazan https://github.com/symfony/symfony/pull/43598
 * [x] Command `cache:pool:clear` (`pools`) @andyexeter https://github.com/symfony/symfony/pull/43621
 * [x] Command `cache:pool:delete` (`pool`, `key`) @andyexeter https://github.com/symfony/symfony/pull/43621
 * [x] Command `completion` (`shell`) @dkarlovi https://github.com/symfony/symfony/pull/43615
 * [x] Command `config:dump-reference` (`name`, `path`) @StaffNowa https://github.com/symfony/symfony/pull/43682
 * [x] Command `help` (`command_name`, `--format`) @GromNaN https://github.com/symfony/symfony/pull/43596
 * [x] Command `debug:translation` (`locale`, `bundle`, `--domain`) @alexandre-daubois https://github.com/symfony/symfony/pull/43644
 * [x] Command `debug:twig` (`name`, `--filter`, `--format`) @StaffNowa https://github.com/symfony/symfony/pull/43846
 * [x] Command `list` (`namespace`, `--format`) @GromNaN https://github.com/symfony/symfony/pull/43596
 * [x] Command `lint:twig` (`filename`, `--format`) @makraz https://github.com/symfony/symfony/pull/43680
 * [x] Command `lint:xliff` (`filename`, `--format`) @makraz https://github.com/symfony/symfony/pull/43680
 * [x] Command `lint:yaml` (`filename`, `--format`, `--exclude`) @makraz https://github.com/symfony/symfony/pull/43680
 * [x] Command `messenger:setup-transports` (`transport`) @Tayfun74 https://github.com/symfony/symfony/pull/43640
 * [x] Command `secrets:remove` (`name`) @GromNaN https://github.com/symfony/symfony/pull/43626
 * [x] Command `secrets:set` (`name`, `file`) @GromNaN https://github.com/symfony/symfony/pull/43626
 * [x] Command `security:encode-password` (`user-class`) @94noni https://github.com/symfony/symfony/pull/43653
 * [x] Command `security:hash-password` (`user-class`) @94noni https://github.com/symfony/symfony/pull/43653
 * [x] Command `server:dump` (`--format`) @alexandre-daubois https://github.com/symfony/symfony/pull/43683
 * [x] Command `translation:pull` (`provider`, `--domains`, `--locales`, `--format`) @welcoMattic https://github.com/symfony/symfony/pull/43672 
 * [x] Command `translation:push` (`provider`, `--domains`, `--locales`) @welcoMattic https://github.com/symfony/symfony/pull/43672 
 * [x] Command `translation:update` (`locale`, `bundle`, `--format`, `--domain`, `--sort`) @stephenkhoo https://github.com/symfony/symfony/pull/43676
 * [x] Command `ulid:generate` (`--format`) @StaffNowa https://github.com/symfony/symfony/pull/43639
 * [x] Command `uuid:generate` (`--format`) @StaffNowa https://github.com/symfony/symfony/pull/43639
 * [x] Command `workflow:dump` (`name`, `marking`, `--label`, `--format`) @StaffNowa StaffNowa https://github.com/symfony/symfony/pull/43848

Commands that does not require auto-completion:

* [x] Command `about`
* [x] Command `assets:install` (`target`) (directory completion)
* [x] Command `cache:clear`
* [x] Command `cache:pool:list`
* [x] Command `cache:pool:prune`
* [x] Command `cache:warmup`
* [x] Command `debug:dotenv`
* [x] Command `dotenv:dump`
 * [x] Command `debug:validator` (`class`) @Kocal *completion support not yet possible*
* [x] Command `lint:container`
* [x] Command `messenger:stop-workers`
* [x] Command `router:match` (`path_info`)
* [x] Command `secrets:decrypt-to-local`
* [x] Command `secrets:encrypt-from-local`
* [x] Command `secrets:generate-keys`
* [x] Command `secrets:list`
* [x] Command `server:log` (`--host`, `--format`) @StaffNowa https://github.com/symfony/symfony/pull/43892
* [x] Command `ulid:inspect`
* [x] Command `uuid:inspect`

**Comment by GromNaN** [comment]
I'm taking `help` and `list` commands.

**Comment by GromNaN** [comment]
1. Since `compgen` filters the suggestion with the current input, can we agree that we don't have to filter in PHP? If filtering is necessary, it should be implemented in `CompletionSuggestions`.

2. Several commands uses the `DescriptorHelper` and have common options `--format`. Some code might be factorized.

**Comment by andyexeter** [comment]
I'll take `cache:pool:clear` and `cache:pool:delete`.

Should I wait for #43596 to be merged? Seems like that has a couple of useful things for testing.

**Comment by GromNaN** [comment]
@andyexeter they are yours.
I will wait for reviews to decide if [the testing class](https://github.com/symfony/symfony/pull/43596/files#diff-daa333a1679e670f76cd7f0e81752cf91f181303608740238353f2269c57c46eR24) is accepted.

**Comment by StaffNowa** [comment]
I can try help uuid and config 🤔

**Comment by IonBazan** [comment]
Can I take following?
- `debug:firewall`
- `debug:form`
- `debug:messenger`
- `debug:router`

Also, is there a way to suggest directories or files? Using Finder for that seems a bit overkill.

**Comment by GromNaN** [comment]
@StaffNowa `ulid:generate` and `uuid:generate` are for you.
@IonBazan you get `debug:firewall`, `debug:form`, `debug:messenger`, `debug:router`.

File and directory autocompletion seems to be a native feature of `compgen`: https://unix.stackexchange.com/a/26733

**Comment by dkarlovi** [comment]
@wouterj and I discussed files/folders being completed by falling back when nothing is suggested by Symfony. It will require a small tweak to the completion script generator.

**Comment by IonBazan** [comment]
@GromNaN I'm asking because `compgen` supports filtering files and directories: https://www.linuxjournal.com/content/more-using-bash-complete-command and some commands we use, like `lint:twig` should only accept `*.twig` files and some other accept only directories, not files.

**Comment by dkarlovi** [comment]
@IonBazan you should assume paths will get completed by `compgen` indeed. :+1: 

The plan is to add a feature to the completion which will allow "signaling" from the completion output to the completion script, something like suggesting (just an example)
```
type: file, glob: *.twig
```

and the completion script running native filtering. I guess it's too late to add this now (feature freeze)? /cc @wouterj

**Comment by fabpot** [comment]
@dkarlovi It's not too late, but we would need to add support quickly if we want to have it in 5.4.

**Comment by dkarlovi** [comment]
@fabpot this described feature is sure nice, but take into account that the completion script should get tweaked that it falls back to the native completion anyway if Symfony doesn't suggest anything, as noted in https://github.com/symfony/symfony/issues/43594#issuecomment-947378122 That is a much smaller change so doable today.

It will result in paths being completed as they usually would, so it goes:
1. use whatever Symfony suggested
2. Symfony didn't suggest anything, use the native file system suggestions

(currently missing 2)

With the signaling feature:
1. use whatever Symfony suggested
1.1. if a signal, use the signal to invoke `compgen` (maybe other things in the future?) in a specific way and use its suggestions
1.2. else a literal value, suggest those
2. Symfony didn't suggest anything, use the native file system suggestions

(1.1. is the new signaling feature)

The "signaling" feature is absolutely a nice one, but IMO not a blocker. If somebody did it in time, it would be awesome, but it should get some design time since the use cases will need to be get worked out. The "fallback to filesystem when nothing suggested" thing is a blocker IMO, but luckily much simpler so can be done before 5.4.

**Comment by wouterj** [comment]
To keep things focused: please open a poc PR or issue if we need to discuss the file completion more in detail. Personally, given the late timing of this feature, I would propose to just rely on the default file autocompletion as a fallback. Then, in 6.1 (or later, whenever someone has time) we can introduce filtered file autocompletion by using signaling.

Symfony <5.4 had zero auto completion, having a "working but not 100% feature complete" autocompletion in 5.4/6.0 is fine.

> Since compgen filters the suggestion with the current input, can we agree that we don't have to filter in PHP? If filtering is necessary, it should be implemented in CompletionSuggestions.

Yes, agreed. We don't have to filter, compgen is capable of that.

**Comment by dkarlovi** [comment]
I just realized falling back when Symfony suggests **nothing** doesn't work because Symfony might be right.

For example, completing this:
```
bin/console compo<TAB>
```

to 

```
bin/console composer. # json or lock
```

is incorrect since **Symfony was right** to suggest nothing: in this context it's completing command names and there is no command matching this pattern, it would yield bad UX.

So, seems signaling (however primitive) is the only correct way forward.

Maybe we can suggest
```
compgen! # using ! as a special character, any syntax can do
```

when we want native completion. Later more semantics can be added.

**Comment by wouterj** [comment]
Taking inspiration from https://github.com/posener/complete/tree/master , the only thing we might need to introduce in 5.4 is a `complete.nothing` "signal".

1. Values suggested? -> complete suggested values
2. "nothing" returned? -> don't do anything
3. No values suggested (e.g. no complete support yet)? -> fallback to default bash behavior (i.e. complete files/dirs)

**Comment by dkarlovi** [comment]
@GromNaN `completion` command actually has an argument which you can suggest, `shell`.

**Comment by welcoMattic** [comment]
@GromNaN Great initiative! I can help to make auto-complete for `translation:pull` and `translation:push` commands

**Comment by GromNaN** [comment]
@welcoMattic you are the one for `translation:pull` and `translation:push`.

**Comment by wouterj** [comment]
Just a recap of the debugging info (as it's not yet documented):

 1. You can enable debug mode by running `export SYMFONY_COMPLETION_DEBUG=1` once in your shell
  2. Then open another terminal window and run `bin/console completion --debug`.
  3. Enjoy live debugging information when trying to use the bash completion in the shell of (1):
     ![image](https://user-images.githubusercontent.com/749025/131250086-1178da81-16ea-410e-9d19-27d63eaad5c7.png)

**Comment by makraz** [comment]
@GromNaN I'm taking `lint` commands

**Comment by dkarlovi** [comment]
@GromNaN I'll take the `completion` command.

**Comment by GromNaN** [comment]
@makraz `lint:twig`, `lint:xliff` and `lint:yaml` are for you.
@dkarlovi `completion` is for you.

**Comment by alexandre-daubois** [comment]
Hey! I'd like to take `debug:translation`! :+1:

**Comment by alexandre-daubois** [comment]
Hey @GromNaN! I'd like to take `debug:translation`! :+1:

**Comment by GromNaN** [comment]
Sure @alexandre-daubois `debug:translation` is in your hands.

**Comment by derrabus** [comment]
If I were to enable commands in 3rd party packages to provide completion, is there a good way to do that if that package should remain compatible with Symfony 4.4? After all, the interface I need to implement does not exist there.

**Comment by dkarlovi** [comment]
@derrabus IMO that should be discussed in a separate issue since it might be of interest to many other people and is likely to be somewhat elaborate.

**Comment by derrabus** [comment]
Agreed. I've created a new issue for that. Sorry for the disruption. 🙂

**Comment by scyzoryck** [comment]
Can I grab following commands, please? :) 

-  `messenger:failed:remove (id, --transport)`
-  `messenger:failed:retry (id, --transport)`
-  `messenger:failed:show (id, --transport)`

**Comment by GromNaN** [comment]
Yes of course @scyzoryck. That's noted.

**Comment by eclairia** [comment]
Hello @GromNaN,

Can I take the following commands?

- `debug:autowiring (search)`
- `debug:config (name, path)`
- `debug:container (name)`

**Comment by wouterj** [comment]
@eclairia sure, they are yours now!

**Comment by Tayfun74** [comment]
Hi, can I take the `messenger:setup-transports` command?

**Comment by GromNaN** [comment]
@Tayfun74 sure, you got it.

**Comment by GromNaN** [comment]
Thank you all for working on the commands. Pull Requests must be rebased since the interface have been removed in #43628.

**Comment by alexandre-daubois** [comment]
Hi @GromNaN, I'd like to take `server:dump` as well! :slightly_smiling_face: 

Also, does the `assets:install` commands really needs completion on `target` argument, as this argument must be a folder?

**Comment by GromNaN** [comment]
Ok @alexandre-daubois 👍🏻

**Comment by 94noni** [comment]
@GromNaN @wouterj Hi,

May I take and work on both commands `security:encode-password (user-class)` & `security:hash-password (user-class)`?
Thanks

**Comment by wouterj** [comment]
Of course @94noni. Good luck :) (I'm not sure if there is a way to find all user classes in an application)

**Comment by 94noni** [comment]
@wouterj I saw that both commands have `private $userClasses`, thus I thought we are talking about this data
I do not understand your comment here, I'll open a PR to discuss on it

**Comment by Kocal** [comment]
Hi everyone, can I try to add autocompletion to `debug:validator` command?

Thanks!

**Comment by wouterj** [comment]
Yes, it's yours now @Kocal

**Comment by Kocal** [comment]
What is the best way to suggest values for `class`? 

Ideally I would like to suggest all classes that can be debugged (ie. with call to `$this->validator->hasMetadaFor($class)`), but how can I get initial classes list: 
- I can't get those classes from the Container (and that's a good thing, since Doctrine Entities are generally excluded from the Container). 
- I can't use `get_declared_classes()` because classes auto-loaded by Composer won't be listed
- I can't use Composer Loader `classMap()`, IFAIK we can't get the ComposerLoader so easily ... and not every projects use Composer

Do you have an idea of how to do that @wouterj?

Thanks!

**Comment by StaffNowa** [comment]
Maybe I can help more 🤔

Command config:dump-reference (name, path)

**Comment by GromNaN** [comment]
@StaffNowa this is one more for you.

**Comment by stephenkhoo** [comment]
I would like to contribute for 
debug:event-dispatcher (event, --dispatcher, --format) and translation:update (locale, bundle, --format, --domain, --sort)

but I have some question on the debug:event-dispatcher
The event seems like have dependency on the --dispatcher selected, will the autocomplete need to take this in account?

**Comment by GromNaN** [comment]
@stephenkhoo `debug:event-dispatcher` and `translation:update` are for you.

You should use already typed options/arguments to make completion more accurate ; like in #43598.

**Comment by stephenkhoo** [comment]
Okay, will take note on that! Thank you.

**Comment by makraz** [comment]
@GromNaN I'm taking commands:
- messenger:consume (receivers, --bus, --queues)
- server:log (--host, --format)

**Comment by wouterj** [comment]
Hi @Kocal! I'm afraid we might not have a way to do that. Completing all class names wouldn't also be that useful (99% of them probably won't have any validation logic), but finishing half a string (e.g. `App\\Dto\\Pr<TAB>`) would be nice, but similarly impossible if we can't get all available class names.

**Comment by Kocal** [comment]
Hi @wouterj, thanks for your response!

That what I thought, even if we filter all available class names with `$validator#hasMetadata()`... If we can't get all available class names, then we can't do anything 😬

So what should we do about `debug:validator` command?

**Comment by StaffNowa** [comment]
I can help with debug:twig 😏

**Comment by dkarlovi** [comment]
Since the release is very close, IMO it's fair to either finish the PRs you've signed up for ASAP or honestly say you'll not make it, maybe somebody else does it in the meantime.

**Comment by fabpot** [comment]
I've made 3 lists in the description to better see which commands have been claimed without PRs yet, commands with a not-yet-merged PR, and commands for which the PR has been merged.

**Comment by wouterj** [comment]
Hi @eclairia @stephenkhoo @StaffNowa @Kocal @makraz @lyrixx. Do you have time to create a PR? It's totally not an issue if you no longer can't - it's voluntary work and your participation in this issue is already much appreciated :pray: - but we would like to finish up Symfony 5.4/6.0 soon. I would love to have completion support for all commands, so maybe someone else can take over/help you :)

**Comment by Kocal** [comment]
Hi @wouterj!

I would like to open a PR to add auto-completion to `debug:validator` but it will empty. 😬 

As said in my previous comments (https://github.com/symfony/symfony/issues/43594#issuecomment-949646700 / https://github.com/symfony/symfony/issues/43594#issuecomment-950295974), I have no idea how we can get all the classes from a project (without dependening on Composer or the Symfony container) 😕 

If someone can take over the feature, that would be appreciated, thanks! 🙏

**Comment by wouterj** [comment]
Ooh, sorry. I went a bit too much into "full automation" mode. Yes, indeed. We decided `debug:validator` can't be done for 5.4/6.0.

**Comment by lyrixx** [comment]
Oups I forgot this issue. And I don't have access to a computer ATM. I'm really sorry 😞. Can someone take over me for the workflow components? Thanks

**Comment by StaffNowa** [comment]
> Hi @eclairia @stephenkhoo @StaffNowa @Kocal @makraz @lyrixx. Do you have time to create a PR? It's totally not an issue if you no longer can't - it's voluntary work and your participation in this issue is already much appreciated pray - but we would like to finish up Symfony 5.4/6.0 soon. I would love to have completion support for all commands, so maybe someone else can take over/help you :)

Hey, I had small business trip :) Return back and made PR request ;)

**Comment by StaffNowa** [comment]
Question about workflow:dump. How we can make autocomplete names from container list? :)

**Comment by 94noni** [comment]
Perhaps you can inject the workflow registry which has access to all workflows/statemachines

**Comment by StaffNowa** [comment]
> Perhaps you can inject the workflow registry which has access to all workflows/statemachines

Found information from $container->getServiceIds(). Very strange, but we do not have any tests at all for workflow:dump. Maybe is it not testable? :)

**Comment by stephenkhoo** [comment]
Hi @wouterj 
I've quite some idea how to implement on the debug:event-dispatcher already, but the debug:event-dispatcher seems doesn't already have a test, do I need to include that test?

I think without the base test, it will be quite hard for me to continue working on with full confident.

Or if I can just submit a PR on the test for debug:event-dispatcher to kick start?

**Comment by StaffNowa** [comment]
> Hi @wouterj I've quite some idea how to implement on the debug:event-dispatcher already, but the debug:event-dispatcher seems doesn't already have a test, do I need to include that test?
> 
> I think without the base test, it will be quite hard for me to continue working on with full confident.
> 
> Or if I can just submit a PR on the test for debug:event-dispatcher to kick start?

Hey, in my case too no any tests in workflow. I think just make first code changes if all good you can write them later before merge code :)

**Comment by GromNaN** [comment]
Hello @makraz, can I take `messenger:consume`?

Edit: done https://github.com/symfony/symfony/pull/43891

**Comment by StaffNowa** [comment]
Hey @makraz, do you have time or we can help you with server:log too?

**Comment by StaffNowa** [comment]
@GromNaN now I see server:log. Do we need to autocomplete hostname and format? According me it is not predictable
--host localhost or any IP (can put some localhost IP addreses).
--format - not sure what we can predict :)

**Comment by GromNaN** [comment]
> @GromNaN now I see server:log. Do we need to autocomplete hostname and format? According me it is not predictable --host localhost or any IP (can put some localhost IP addreses). --format - not sure what we can predict :)

What do you think of suggesting the default value for both? It would be useful as a starting point, instead of reading the doc.

**Comment by StaffNowa** [comment]
> > @GromNaN now I see server:log. Do we need to autocomplete hostname and format? According me it is not predictable --host localhost or any IP (can put some localhost IP addreses). --format - not sure what we can predict :)
> 
> What do you think of suggesting the default value for both? It would be useful as a starting point, instead of reading the doc.

Yes good point👌

**Comment by StaffNowa** [comment]
@GromNaN added default hosts and ports from real life :)

**Comment by wouterj** [comment]
Hi @stephenkhoo 

> do I need to include that test?

Yes, you can add a test for the completion support.

There is no direct need to add all missing tests (but if you want to do that as well, a PR is welcome to e.g. 4.4 or 5.3).

**Comment by makraz** [comment]
> Hey @makraz, do you have time or we can help you with server:log too?

Hi @StaffNowa sorry just back to online work

**Comment by StaffNowa** [comment]
@makraz Hey, it is not needed anymore for the server:log command. People decided do not nothing for this command :)
